### PR TITLE
Support null task affinity for activities used to launch interactive token requests

### PR DESCRIPTION
**Pull request by shoatman**
When authentication is completed interactive token requests that were started from an activity which do not have a taskAffinity (null taskAffinity) that activity is not brought to the foreground after the interaction with the authorization endpoint is complete.  This is a problem for some apps that leverage null taskaffinity in multi-window scenarios.

### CODE REVIEW

**Comment by shoatman** [comment]
This needs changelog information as well as likely a documentation update for configuration as well.

File: common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationStrategy.java [hunk 1]
```diff
@@ -50,6 +51,7 @@
     private WeakReference<Context> mReferencedApplicationContext;
     private WeakReference<Activity> mReferencedActivity;
     private WeakReference<Fragment> mReferencedFragment;
+    protected int mTaskIdThatStartedAuthorization;
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationStrategy.java:54]: Whoops i didn't end up using this.

File: common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java [hunk 2]
```diff
@@ -429,6 +431,38 @@ public void run() {
         });
     }
 
+
+    /**
+     * This method optionally re-orders that task to bring the task that launched
+     * the interactive activity to the foreground.  This is useful when the activity provided
+     * to us does not have a taskAffinity and as a result it's possible that other apps or the home
+     * screen could be in the task stack ahead of the app that launched the interacxtive
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/internal/controllers/CommandDispatcher.java:439]: typo

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java [hunk 3]
```diff
@@ -39,21 +43,50 @@
...
+        final PackageManager packageManager = applicationContext.getPackageManager();
+        try {
+            ActivityInfo startActivityInfo = packageManager.getActivityInfo(parameters.getActivity().getComponentName(), 0);
+            if(startActivityInfo.taskAffinity == null){
+                mHasTaskAffinity = false;
+                mTaskId = parameters.getActivity().getTaskId();
+            }
+        } catch (PackageManager.NameNotFoundException e) {
+            //TODO: Log properly and then ignore
+            e.printStackTrace();
```
Comment (shoatman) [inline common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:77]: Add correct logging.

File: changelog.txt [hunk 4]
```diff
@@ -1,6 +1,7 @@
 V.Next
 ----------
 - [MINOR] Enables removeAccount api to remove account records from all environments (#1248)
+- [MINOR] Added support for handling null taskAffinity.  InteractiveTokenCommmand now captures whether taskAFfinity is null and records taskId
```
Comment (iambmelt) [inline changelog.txt:4]:
```suggestion
- [MINOR] Added support for handling null taskAffinity.  InteractiveTokenCommmand now captures whether taskAFfinity is null and records taskId (#1282)
```

File: common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationStrategy.java [hunk 5]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/providers/oauth2/AuthorizationStrategy.java:1]: stray newline

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java [hunk 6]
```diff
@@ -39,21 +43,53 @@
...
 
     public InteractiveTokenCommand(@NonNull InteractiveTokenCommandParameters parameters,
                                    @NonNull List<BaseController> controllers,
                                    @SuppressWarnings(WarningType.rawtype_warning) @NonNull CommandCallback callback,
                                    @NonNull final String publicApiId) {
         super(parameters, controllers, callback, publicApiId);
+        checkAndRecordTaskInformation(parameters);
     }
 
+    private void checkAndRecordTaskInformation(InteractiveTokenCommandParameters parameters){
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:66]:
```suggestion
    private void checkAndRecordTaskInformation(@NonNull final InteractiveTokenCommandParameters parameters){
```

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java [hunk 7]
```diff
@@ -39,21 +43,53 @@
...
+        final String methodName = ":checkAndRecordTaskInformation";
+        final Context applicationContext = parameters.getAndroidApplicationContext();
+        final PackageManager packageManager = applicationContext.getPackageManager();
+        try {
+            ActivityInfo startActivityInfo = packageManager.getActivityInfo(parameters.getActivity().getComponentName(), 0);
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:71]: IIRC, OneAuth added some scenarios where the calling `Activity` might be `null`, we may want to be defensive here and check it's `!= null` before requesting the `componentName`

/cc @irving-as

Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:71]:
```suggestion
            final ActivityInfo startActivityInfo = packageManager.getActivityInfo(parameters.getActivity().getComponentName(), 0);
```

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java [hunk 8]
```diff
@@ -39,21 +43,53 @@
...
+            if(startActivityInfo.taskAffinity == null){
+                mHasTaskAffinity = false;
+                mTaskId = parameters.getActivity().getTaskId();
+            }
+        } catch (PackageManager.NameNotFoundException e) {
```
Comment (iambmelt) [inline common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:76]:
```suggestion
        } catch (final PackageManager.NameNotFoundException e) {
```

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java (hunk 7 above)
Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java]: Not sure how that works... when would activity be null here?

Comment (iambmelt) [reply common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java]: @irving-as Something you can comment on? As I recall, there is behavior where the broker / interactive flows will be invoked using only an `Activity` context

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java [hunk 9]
```diff
@@ -39,21 +43,53 @@
...
+            }
+        } catch (final PackageManager.NameNotFoundException e) {
+            Logger.warn(
+                    TAG + methodName,
+                    "Unable to get ActivityInfo for activity provided to start authorization."
+            );
+        }
+    }
+
+    public boolean getHasTaskAffinity(){
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:84]:
```suggestion
    public boolean hasTaskAffinity(){
```

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java [hunk 10]
```diff
@@ -39,21 +43,53 @@
 @EqualsAndHashCode(callSuper = true)
 public class InteractiveTokenCommand extends TokenCommand {
     private static final String TAG = InteractiveTokenCommand.class.getSimpleName();
+    //Normally all tasks have an affinity unless configured explicitly for multi-window support to not have one
+    private boolean mHasTaskAffinity = true;
+    private int mTaskId = 0;
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:48]: nit:

@Getter 
@Accessors(prefix = "m")

?

Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:48]: oh boy... let's take this up with Brian and the rest of the team.  I prefer not to do this myself... however it's part of the android naming guidelines

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java (hunk 9 above)
Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:84]: To be honest I don't care what the standard is... as long as we're all following it.  Unfortunately we have a diversity of things now in the code.

File: common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java (hunk 10 above)
Comment (AdamBJohnsonx) [reply common/src/main/java/com/microsoft/identity/common/internal/commands/InteractiveTokenCommand.java:48]: I think you misunderstand.  Putting those two annotations on this class will create the accessors understanding that the member prefix is "m", and name the methods without it.

**Review by AdamBJohnsonx — APPROVED** [review]
I don't have problems here.

**Pull request merged**